src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+                    // the initializer. Validate the escape values for the initializer instead.
+
+                    Debug.Assert(localSymbol.RefKind == RefKind.None ||
+                        refEscapeScope >= GetRefEscape(initializer, _localScopeDepth));
+
+                    if (node.DeclaredTypeOpt?.Type.IsRefLikeType == true)
+                    {
+                        ValidateEscape(initializer, valEscapeScope, isByRef: false, _diagnostics);
+                    }
+                }
+                else
+                {
+                    // default to the current scope in case we need to handle self-referential error cases.
+                    SetLocalScopes(localSymbol, _localScopeDepth, _localScopeDepth);
+
+                    valEscapeScope = GetValEscape(initializer, _localScopeDepth);
+                    if (localSymbol.RefKind != RefKind.None)
+                    {
+                        refEscapeScope = GetRefEscape(initializer, _localScopeDepth);
+                    }
+
+                    SetLocalScopes(localSymbol, refEscapeScope, valEscapeScope);
+                }
+            }
+
+            return null;
+        }
+
+        public override BoundNode? VisitReturnStatement(BoundReturnStatement node)
+        {
+            base.VisitReturnStatement(node);
+            if (node.ExpressionOpt is { Type: { } } expr)
+            {
+                ValidateEscape(expr, Binder.ReturnOnlyScope, node.RefKind != RefKind.None, _diagnostics);
+            }
+            return null;
+        }
+
+        public override BoundNode? VisitYieldReturnStatement(BoundYieldReturnStatement node)
+        {
+            base.VisitYieldReturnStatement(node);
+            if (node.Expression is { Type: { } } expr)
+            {
+                ValidateEscape(expr, Binder.ReturnOnlyScope, isByRef: false, _diagnostics);
+            }
+            return null;
+        }
+
+        public override BoundNode? VisitAssignmentOperator(BoundAssignmentOperator node)
+        {
+            base.VisitAssignmentOperator(node);
+            if (node.Left.Kind != BoundKind.DiscardExpression)
+            {
+                ValidateAssignment(node.Syntax, node.Left, node.Right, node.IsRef, _diagnostics);
+            }
+            return null;
+        }
+
+        public override BoundNode? VisitIsPatternExpression(BoundIsPatternExpression node)
+        {
+            using var _ = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            return base.VisitIsPatternExpression(node);
+        }
+
+        public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+        {
+            SetLocalScopes(node);
+
+            using var _ = new PatternInput(this, getDeclarationValEscape(node.DeclaredType, _patternInputValEscape));
+            return base.VisitDeclarationPattern(node);
+
+            static uint getDeclarationValEscape(BoundTypeExpression typeExpression, uint valEscape)
```

Comment:
nit: consider using a local instead of a local function (seems simpler): `var declarationValEscape = node.DeclaredType.Type.IsRefLikeType ? _patternInputValEscape: Binder.CallingMethodScope;`

Reply:
A similar local function is used in `VisitPositionalSubpattern` (as you mentioned) and in `VisitPropertySubpattern`, and the local function in `VisitPropertySubpattern` is more difficult to inline because it is recursive. I'll leave these as is, for consistency.